pkg/allocateip/run_test.go

```diff
@@ -0,0 +1,201 @@
+			Expect(n.Spec.BGP.IPv4IPIPTunnelAddr).ToNot(Equal(""))
+		})
+	})
+
+	Context("VXLAN tunnel address tests", func() {
+		It("should add tunnel address to node", func() {
+			node := makeNode("192.168.0.1/24", "fdff:ffff:ffff:ffff:ffff::/80")
+			node.Name = "test.node"
+
+			_, err := c.Nodes().Create(ctx, node, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			_, ip4net, _ := net.ParseCIDR("172.16.0.0/24")
+			ensureHostTunnelAddress(ctx, c, node.Name, []net.IPNet{*ip4net}, true)
+			n, err := c.Nodes().Get(ctx, node.Name, options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(n.Spec.IPv4VXLANTunnelAddr).ToNot(Equal(""))
+		})
+
+		It("should add tunnel address to node without BGP Spec", func() {
+			node := makeNode("192.168.0.1/24", "fdff:ffff:ffff:ffff:ffff::/80")
+			node.Name = "test.node"
+			node.Spec.BGP = nil
+
+			_, err := c.Nodes().Create(ctx, node, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			_, ip4net, _ := net.ParseCIDR("172.16.0.0/24")
+			ensureHostTunnelAddress(ctx, c, node.Name, []net.IPNet{*ip4net}, true)
+			n, err := c.Nodes().Get(ctx, node.Name, options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(n.Spec.IPv4VXLANTunnelAddr).ToNot(Equal(""))
+		})
+
+	})
+})
+
+var _ = Describe("removeHostTunnelAddr", func() {
+	log.SetOutput(os.Stdout)
+	// Set log formatting.
+	log.SetFormatter(&logutils.Formatter{})
+	// Install a hook that adds file and line number information.
+	log.AddHook(&logutils.ContextHook{})
+
+	ctx := context.Background()
+	cfg, _ := apiconfig.LoadClientConfigFromEnvironment()
+
+	var c client.Interface
+	BeforeEach(func() {
+		// Clear out datastore
+		be, err := backend.NewClient(*cfg)
+		Expect(err).ToNot(HaveOccurred())
+		be.Clean()
+
+		//create client and IPPool
+		c, _ = client.New(*cfg)
+		c.IPPools().Create(ctx, makeIPv4Pool("172.16.0.0/24"), options.SetOptions{})
+	})
+
+	Context("IPIP tunnel address tests", func() {
+		It("should remove tunnel address from node", func() {
+			node := makeNode("192.168.0.1/24", "fdff:ffff:ffff:ffff:ffff::/80")
+			node.Name = "test.node"
+			node.Spec.BGP.IPv4IPIPTunnelAddr = "172.16.0.5"
+
+			_, err := c.Nodes().Create(ctx, node, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			removeHostTunnelAddr(ctx, c, node.Name, false)
+			n, err := c.Nodes().Get(ctx, node.Name, options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(n.Spec.BGP.IPv4IPIPTunnelAddr).To(Equal(""))
+		})
+
+		It("should not panic on node without BGP Spec", func() {
+			node := makeNode("192.168.0.1/24", "fdff:ffff:ffff:ffff:ffff::/80")
+			node.Name = "test.node"
+			node.Spec.BGP = nil
+
+			_, err := c.Nodes().Create(ctx, node, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			removeHostTunnelAddr(ctx, c, node.Name, false)
+			n, err := c.Nodes().Get(ctx, node.Name, options.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(n.Spec.BGP).To(BeNil())
+		})
+	})
+
```

Comment:
Worth checking for cross-talk?  I.e. the vxlan code shouldn't remove an IPIP address and vice versa